src/NHibernate/Impl/ExpressionProcessor.cs

```diff
@@ -248,32 +248,78 @@ private static ICriterion Le(ProjectionInfo property, object value)
-				return ((ConstantExpression) expression).Value;
-
-			if (expression.NodeType == ExpressionType.MemberAccess)
+			object value;
+			switch (expression.NodeType)
 			{
-				var memberAccess = (MemberExpression) expression;
-				if (memberAccess.Expression == null || memberAccess.Expression.NodeType == ExpressionType.Constant)
-				{
-					var constantValue = ((ConstantExpression) memberAccess.Expression)?.Value;
-					var member = memberAccess.Member;
-					switch (member.MemberType)
+				case ExpressionType.Constant:
+					var constantExpression = (ConstantExpression) expression;
+					return constantExpression.Value;
+				case ExpressionType.MemberAccess:
+					var memberExpression = (MemberExpression) expression;
+					value = memberExpression.Expression != null ? FindValue(memberExpression.Expression) : null;
+
+					switch (memberExpression.Member.MemberType)
 					{
 						case MemberTypes.Field:
-							return ((FieldInfo) member).GetValue(constantValue);
+							return ((FieldInfo) memberExpression.Member).GetValue(value);
 						case MemberTypes.Property:
-							return ((PropertyInfo) member).GetValue(constantValue);
+							return ((PropertyInfo) memberExpression.Member).GetValue(value);
 					}
-				}
+					break;
+				case ExpressionType.Call:
+					var methodCallExpression = (MethodCallExpression) expression;
+					var args = new object[methodCallExpression.Arguments.Count];
+					for (int i = 0; i < args.Length; i++)
+						args[i] = FindValue(methodCallExpression.Arguments[i]);
+
+					if (methodCallExpression.Object == null) //extension or static method
+					{
+						if (args.Length > 0 && args[0] != null)
+						{
+							return methodCallExpression.Method.Invoke(args[0].GetType(), args);
+						}
+						else
+						{
+							return methodCallExpression.Method.Invoke(methodCallExpression.Method.DeclaringType, args);
+						}
+					}
+					else
+					{
+						var callingObject = FindValue(methodCallExpression.Object);
+
+						return methodCallExpression.Method.Invoke(callingObject, args);
+					}
+				case ExpressionType.Convert:
+					var unaryExpression = (UnaryExpression) expression;
+					value = FindValue(unaryExpression.Operand);
+					var type = Nullable.GetUnderlyingType(unaryExpression.Type) ?? unaryExpression.Type;
+					if (type == typeof(object) || value == null)
+					{
+						return value;
+					}
+					else if (value is IConvertible || unaryExpression.Method != null)
+					{
+						if (type != unaryExpression.Operand.Type)
+						{
+							value = unaryExpression.Method != null ? unaryExpression.Method.Invoke(null, new[] { value }) : Convert.ChangeType(value, type);
+						}
+
+						return value;
+					}
+					break;
 			}
 
-			var valueExpression = Expression.Lambda(expression).Compile();
-			object value = valueExpression.DynamicInvoke();
+			var valueExpression = Expression.Lambda(expression)
+#if NETCOREAPP2_0_OR_GREATER || NET47_OR_GREATER || NETSTANDARD2_0_OR_GREATER
```

Comment:
Please create an extension method instead:

```csharp
#if NET461
internal static class LambdaExpressionExtensions 
{
  public Delegate Compile (this LambdaExpression expression, bool preferInterpretation) =>
    expression.Compile();    
}
#endif
```

Reply:
This means a net47-48 application cannot benefit from this optimization until we upgrade the NHibernate target to net47, right? (Or add a target for net48 by example.)

Reply:
I think community needs to discuss, to apply this fix to older versions of nhibernate for fixing GC issue.

Reply:
@fredericDelaporte it is essentially the same code just shifted to an extension method to reduce noise. I did not do any thinking beyond that. We probably can build a runtime shim to discover the underlying method (as we do for other stuff).

Reply:
If we are going to add net60 target to 5.4 than why not also add net47 (or net48 but seems net48 doesn't add anything useful for us). I would prefer to add target rather than adding such shims.

Reply:
MS recommends adding 4.8

Reply:
It seems preferInterpretation is a lie under .NET Framework:
https://github.com/microsoft/referencesource/blob/NetFramework48ZDP/System.Core/Microsoft/Scripting/Ast/LambdaExpression.cs#L149